refactor!: Server Script (Scheduler Event) & misc APIs

gavindsouza opened the pull request:
### Changes

 - Disable the previous "Scheduled Job Type" document instead of deleting it
 - Document.has_value_changed returns Truthy or False 
 - ~Expose safe_exec as a ServerScript namespace method~
 - Use cached ServerScript doc for executing doc events

gavindsouza committed:
perf: Use cached Server Script to execute doc method events

gavindsouza committed:
fix: Set value default True for Truthy cls

gavindsouza committed:
fix!: Dont delete existing ScheduledJobType on change

Prior to this, any change in the cron would lead to previous SJT
documents & corresponding SJL documents deleted which may be
undesirable.

gavindsouza committed:
fix!: Document.has_value_changed returns Truthy or False

 - Return changed value to avoid re-accessing previous object & it's
   attribute
 - Wrap returned value as Truthy to avoid breaking change in API

gavindsouza committed:
fix: ServerScript.safe_exec as a doc method

ankush commented:
> Expose safe_exec as a ServerScript namespace method

This isn't acceptable. Unnecessary increase in exposure of surface. Also https://github.com/frappe/frappe/issues/26385


> Document.has_value_changed returns Truthy or False

Use enums instead?  If goal is to distinguish between changed value vs non-existing previous value then:

```python
import enum

class ValueChangedResult(enum.Flag):
	ValueChanged = True
	NoPreviousValue = True
	ValueUnchanged = False
```

gavindsouza committed:
Revert "fix: ServerScript.safe_exec as a doc method"

This reverts commit cf2a3e926e92094f77a8355f61ca7cd37290802b.

gavindsouza commented:
> Use enums instead? If goal is to distinguish between changed value vs non-existing previous value then:

My first expectation from the API `has_value_changed` was that it would return the previous value (similar to the exists API et al). But that would be breaking behaviour since the previously stored value may not be truthy ('', 0) to correctly indicate if the value has been changed.

Currently, to know if a value has changed from, we have to write: 

```py
if doc.has_value_changed("field"):
  previous_doc = doc.get_doc_before_save()
  previous_value_field = previous_doc.field
```

And now that would change to:

```py
if wrapped_previous_value := doc.has_value_changed("field"):
  previous_value_field = wrapped_previous_value.value
```

---

Although this change is not crucial to the PR and can be reverted just like the interfacing safe_exec change, it makes usage more convenient - and there's room for improvement.

ankush commented on frappe/utils/__init__.py:
```diff
@@ -1166,3 +1168,18 @@ def run(self):
+
+	def __bool__(self):
+		return True
+
```
```suggestion

	def __eq__(self, other: object) -> bool:
		return True == other
```

`==` doesn't work. 

`is` too but that isn't fixable.

gavindsouza replied:
> == doesn't work.

I'm assuming we want to implement this for the sake of backwards compatibility where someone is using `doc.has_value_changed("field") == True`. Otherwise, there would be no reason to support this, eh?

ankush replied:
yeah

gavindsouza committed:
fix: Implement Truthy.__eq __

Implement equality check, compare with True rather than self for a Truthy instance

Co-authored-by: Ankush Menat <ankushmenat@gmail.com>
